nuget-sdk-usage/nuget-sdk-usage/AssemblyAnalyser.cs

```diff
@@ -0,0 +1,200 @@
+                                AddMember(assembly, member);
+                            }
+                            break;
+
+                        default:
+                            throw new NotImplementedException(kind.ToString());
+                    }
+                }
+            }
+
+            if (result.MemberReferences.SelectMany(r => r.Value).Any())
+            {
+                if (TryGetTargetFramework(metadata, out string? targetFramework))
+                {
+                    result.TargetFrameworks.Add(targetFramework);
+                }
+            }
+
+            return result;
+        }
+
+        private static string GetMethodSignature(string member, MethodSignature<string> sig)
+        {
+            var sb = new StringBuilder();
+
+            sb.Append(member);
+            sb.Append('(');
+            for (int i = 0; i < sig.ParameterTypes.Length; i++)
+            {
+                if (i != 0)
+                {
+                    sb.Append(", ");
+                }
+
+                sb.Append(sig.ParameterTypes[i]);
+            }
+            sb.Append(')');
+
+            return sb.ToString();
+        }
+
+        private static bool DefinedInNuGetAssembly(AssemblyReference assemblyReference)
+        {
+            var assemblyName = assemblyReference.GetAssemblyName();
+
+            if (assemblyName.Name != null && !NuGetAssembly.MatchesName(assemblyName.Name))
+            {
+                return false;
+            }
+
+            // The strong name keys NuGet uses is in the repo, as is good practise, so we can't be sure
+            // that the 
+            var publicKeyToken = assemblyName.GetPublicKeyToken();
+            if (publicKeyToken == null)
+            {
+                return false;
+            }
+
+            return NuGetAssembly.IsMicrosoftPublicToken(publicKeyToken);
+        }
+
+        private static bool TryFindAssemblyReference(EntityHandle handle, MetadataReader metadata, out AssemblyReference assemblyReference)
+        {
+            assemblyReference = default;
+            switch (handle.Kind)
+            {
+                case HandleKind.AssemblyReference:
+                    {
+                        assemblyReference = metadata.GetAssemblyReference((AssemblyReferenceHandle)handle);
+                        return true;
+                    }
+
+                case HandleKind.TypeReference:
+                    {
+                        var typeReference = metadata.GetTypeReference((TypeReferenceHandle)handle);
+                        return TryFindAssemblyReference(typeReference.ResolutionScope, metadata, out assemblyReference);
+                    }
+
+                case HandleKind.TypeSpecification:
+                    {
+                        // I don't know what this is, or how to find the assembly the type is defined in.
```

Comment:
From my understanding, a `TypeSpecification` is a generic instantiation (like `List<int>`). I believe the `TypeReference`s and `TypeDefinition`s should have the open generic (`List<T>`)

Reply:
Yes, you're right. This is a flaw in me using a signature decoder that decodes straight to a string. During decoding, `ISignatureTypeProvider`'s `GetTypeFromReference` and `GetGenericInstantiation` is called, and I can get the assembly from the `TypeReference`.

A quick, hacky option is to implement a `ISignatureTypeProvider` that returns the assembly name as a string, rather than the class name. I don't know if that's better or worse than pretending that the type wasn't defined in any assembly, which therefore gives wrong results, as I'm doing now. To do it "right", I'd need to build a model that can represent classes, methods, generics, and so on, and the signature type provider would need to return that instead of a string.

I can't think of any generic types that NuGet SDK exposes, so for now I'm going to say I think it's ok.